homework-20-12-23/sos-ghazaryan/cleanup.sh

```diff
@@ -0,0 +1,95 @@
+#!/bin/bash -e
+
+function cleanResourcesByTag
+{
+#  $1 tag name
+#  $2 tag value
+  echo "cleaning resources with tag ${1} and value ${2} ..."
+  if [ -n "${2}" ]
+  then
+   cleanInstancesByTag ${1} ${2}
+   cleanRouteTables ${1} ${2}
+   cleanGateWays ${1} ${2}
+   cleanSubnets ${1} ${2}
+   cleanSecurityGroups ${1} ${2}
+   cleanVPCs ${1} ${2}
+  fi
+  echo 'end of cleaning...'
+}
+
+function cleanInstancesByTag
+{
+  # Get all instances
+  instances=$(aws ec2 describe-instances --query 'Reservations[*].Instances[*]' --output json)
+
+  # Filter instances where tag value is not equal to EXCLUDE_TAG_VALUE
+  instances_to_delete=$(echo $instances | jq -r '.[][] | select(.Tags | any(.Key == "'${1}'" and .Value != "'${2}'")).InstanceId')
+
+  # Loop through each instance and terminate it
+  for instance_id in $instances_to_delete; do
+      echo "Terminating instance $instance_id"
+      aws ec2 terminate-instances --instance-ids $instance_id
```

Comment:
The instance takes time to reach terminated state. No resources can be deleted or detached before the instance is in terminated state. There should be a wait/check if the instance has reached terminated state before deleting the rest of the resources.
Currently the cleanup script works if you run it multiple times, first it shuts down the instance and throws errors about resources being in use, if you want for a minute or two and rerun it, the rest of the resources are deleted.